File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/WeightyImpactEffect.ts
import {AbstractEffect} from "mods/ModTheSpire/Scripts/Effect/AbstractEffect";
import Color from "mods/ModTheSpire/Scripts/DataDefine/Color";
import {FairyGUI} from "csharp";
import UIHelper from "mods/ModTheSpire/Scripts/UI/UIHelper";
import FileHelper from "mods/ModTheSpire/Scripts/FileHelper";
import {Mathf} from "Core/Module/Math/Mathf";
import {TimeKit} from "Core/Utils/TimeKit";
import {SoundMaster} from "mods/ModTheSpire/Scripts/Audio/SoundMaster";
import DungeonManager from "mods/ModTheSpire/Scripts/DungeonManager";
import {BorderFlashEffect} from "mods/ModTheSpire/Scripts/Effect/Combat/BorderFlashEffect";
import {ShakeDur, ShakeIntensity, ShakeScreen} from "mods/ModTheSpire/Scripts/Effect/ShakeScreen";
import {DamageImpactCurvyEffect} from "mods/ModTheSpire/Scripts/Effect/Combat/DamageImpactCurvyEffect";
import {UpgradeShineParticleEffect} from "mods/ModTheSpire/Scripts/Effect/UpgradeShineParticleEffect";
import * as TWEEN from "Core/Module/Tween"
export class WeightyImpactEffect extends AbstractEffect{
    public Loader:FairyGUI.GLoader[] = new FairyGUI.GLoader[3]
    private _TargetY:number;
    private _Scale:number = 1;
    private _ImpactHook :boolean = false;
    private _InitY:number;
    private _InitHeight = 0;
    public constructor(private _X:number,private _Y:number,private _Color?:Color) {
        super();
        this._InitY = FairyGUI.GRoot.inst.height;
        this._TargetY = this._Y - 180;
        for(let i = 0;i<3;i++)
            this.Create(i);
        this._InitHeight = this.Loader[0].height;
        new TWEEN.Tween({y:this._InitY}).to({y:this._TargetY}).easing(TWEEN.Easing.Sinusoidal.InOut).onUpdate(t1=>this._Y = t1.y).start()
    }
    
    private Create(i:number){
        this.Loader[i] = UIHelper.CreateGLoader();
        this.Loader[i].url = FileHelper.FormatPath("Vfx/combat/weightImpact");
        this.Loader[i].SetPivot(0.5,0.5,true);
        this.Loader[i].SetPosition(this._X,this._InitY,0);
        this.Loader[i].rotation = Mathf.RandomRange(-1,1);
        this.Loader[i].color = this._Color.UnityColor();
        this.Loader[i].alpha = 0.6;
        this.Loader[i].TweenFade(0,0.2).SetEase(FairyGUI.EaseType.SineInOut).SetDelay(0.8);
        this.Loader[i].TweenFade(0.2,0.8).SetEase(FairyGUI.EaseType.QuadInOut);
        this.Loader[i].blendMode = FairyGUI.BlendMode.Add;
        this.Loader[i].SetSize(this.Loader[i].texture.width,this.Loader[i].texture.height);
        this.Loader[i].fill = FairyGUI.FillType.ScaleFree;
    }
    
    Update() {
        super.Update();
        this._Scale += TimeKit.DeltaTime;
        this.Loader[0].SetScale(this._Scale* Mathf.RandomRange(0.99, 1.01) * 0.3,this._Scale* Mathf.RandomRange(0.99, 1.01) * 2 * (1 - this.Duration + 0.8));
        this.Loader[1].SetScale(this._Scale* Mathf.RandomRange(0.99, 1.01) * 0.7,this._Scale* Mathf.RandomRange(0.99, 1.01) * 1.3 * (1 - this.Duration + 0.8));
        this.Loader[2].SetScale(this._Scale* Mathf.RandomRange(0.99, 1.01),this._Scale* Mathf.RandomRange(0.99, 1.01) * (1 - this.Duration + 0.8));
        for(let i = 0;i<3;i++){
            this.Loader[i].height = this._InitHeight * (1 - this.Duration + 2) * 5;
        }
        this.Loader[0].y = this._Y + 140;
        this.Loader[1].y = this._Y + 40;
        this.Loader[2].y = this._Y;
        if(this.Duration >= 1){
            this.IsDone = true;
            SoundMaster.PlayVoiceA("ATTACK_IRON_2", -0.5);
            return;
        } else if(this.Duration > 0.8){
            if(this._ImpactHook == false){
                let i = 0;
                this._ImpactHook = true;
                DungeonManager.EffectManager.Play(new BorderFlashEffect(Color.Orange));
                DungeonManager.EffectManager.Play(new ShakeScreen(ShakeIntensity.HIGH,ShakeDur.MED,{Vertical: true,Horizontal: false}));
                for(i = 0;i<10;i++){
                    DungeonManager.EffectManager.Play(new DamageImpactCurvyEffect(this._X,this._Y))
                }
                for(i = 0;i<30;i++){
                    DungeonManager.EffectManager.Play(new UpgradeShineParticleEffect(this._X + Mathf.RandomRange(-100,100),this._Y + Mathf.RandomRange(-50,120)))
                }
            }
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/WeightyImpactEffect.cs
using Core.Utils;
using Core.Utils.Unity;
using DG.Tweening;
using FairyGUI;
using ModTheSpire.DataDefine;
using ModTheSpire.Dungeon;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class WeightyImpactEffect : AbstractEffect
    {
        private float mX;
        private float mY;
        private Color mColor;
        private readonly float mInitY;
        private readonly float mTargetY;
        private FairyGUI.GLoader[] mLoader = new FairyGUI.GLoader[3];
        private float mInitHeight = 0;
        private float mScale;
        private bool mImpactHook;
        public WeightyImpactEffect(float x, float y): this(x,y,new Color(1,1,0.1f,0)) { }
        public WeightyImpactEffect(float x, float y, Color color)
        {
            mX = x;
            mColor = color;

            this.mInitY = FairyGUI.GRoot.inst.height;
            this.mTargetY = y - 180;
            this.Create();
            this.mInitHeight = mLoader[0].height;
            DOTween.To(() => mInitY, t1 => mY = t1, mTargetY, 1).SetEase(Ease.InOutSine);
        }
        
        private void Create()
        {
            for (int i = 0; i < 3; i++)
            {
                mLoader[i] = UIObjectFactory.NewObject(ObjectType.Loader).asLoader;
                mLoader[i].url = "Vfx/combat/weightImpact";
                mLoader[i].SetPivot(0.5f,0.5f,true);
                mLoader[i].SetPosition(this.mX,this.mInitY,0);
                mLoader[i].rotation = Random.Range(-1f,1f);
                mLoader[i].color = mColor;
                mLoader[i].alpha = 0.6f;
                mLoader[i].TweenFade(0,0.2f).SetEase(EaseType.SineInOut).SetDelay(0.8f);
                mLoader[i].TweenFade(0.2f,0.8f).SetEase(EaseType.QuadInOut);
                mLoader[i].blendMode = BlendMode.Add;
                mLoader[i].SetSize(mLoader[i].texture.width,mLoader[i].texture.height);
                mLoader[i].fill = FillType.ScaleFree;
            }
        }

        public override void Update()
        {
            base.Update();
            mScale += Time.deltaTime;
            mLoader[0].SetScale(mScale* Random.Range(0.99f, 1.01f) * 0.3f,mScale* Random.Range(0.99f, 1.01f) * 2 * (1 - this.Duration + 0.8f));
            mLoader[1].SetScale(mScale* Random.Range(0.99f, 1.01f) * 0.7f,mScale* Random.Range(0.99f, 1.01f) * 1.3f * (1 - this.Duration + 0.8f));
            mLoader[2].SetScale(mScale* Random.Range(0.99f, 1.01f),mScale* Random.Range(0.99f, 1.01f) * (1 - this.Duration + 0.8f));
            for(var i = 0;i<3;i++){
                mLoader[i].height = this.mInitHeight * (1 - this.Duration + 2) * 5;
            }
            mLoader[0].y = this.mY + 140;
            mLoader[1].y = this.mY + 40;
            mLoader[2].y = this.mY;
            if(this.Duration >= 1){
                this.IsDone = true;
                SoundMaster.PlayVoice(SoundMaster.ATTACK_IRON_2, -0.5f, false);
            } else if(this.Duration > 0.8){
                if(this.mImpactHook == false){
                    var i = 0;
                    this.mImpactHook = true;
                    DungeonManager.EffectManager.Play(new BorderFlashEffect(MTSColor.Orange.Clone()));
                    CameraUtils.Shake(ShakeVibrato.HIGH, ShakeDur.MED, new Vector2(0, 0.1f));
                    for(i = 0;i<10;i++)
                    {
                        DungeonManager.EffectManager.Play(new DamageImpactCurvyEffect(mX, mY));
                    }
                    for(i = 0;i<30;i++)
                    {
                        DungeonManager.EffectManager.Play(new UpgradeShineParticleEffect(mX + Random.Range(-100, 100), mY + Random.Range(-50, 120)));
                    }
                }
            }
        }
    }
}